### PR TITLE
cli(output): Add ability to export results to CSV

### PR DESCRIPTION
**Pull request by nourikhalass**
These changes implement the ability to export results to an CSV file.
These changes address the idea proposed in #4325 and #4481.

Each row of the CSV file contains the information of one audit. Currently the following properties are exported:

- Name of the category the audit belongs to
- Name of the audit
- Description of the audit
- Score type used for the audit
- Score value

These properties were selected arbitrarily and perhaps there are more / better ones to export. The implementation relies on type information about the result object that is passed to the `Printer` in `lighthouse-cli`. This information was lacking so it has been added in order to satisfy the typechecker. However, the added type information should probably be expanded further and used in other places to have better type safety in the complete code base.

### CODE REVIEW

**Comment by nourikhalass** [comment]
@mvanspaendonck, @anandkanav92, @ernstmul Y'all need to give Google the Good Ol' "[I hereby give up my rights](https://www.youtube.com/watch?v=42FRgdjveGU)"!

**Comment by mvanspaendonck** [comment]
The cla should be signed by me already, but this is still aall fine with me.

**Comment by anandkanav92** [comment]
Fine by me 🤷🏼‍♂️

**Comment by ernstmul** [comment]
Fine by me as well!

File: lighthouse-cli/printer.js [hunk 1]
```diff
@@ -32,6 +34,48 @@ function checkOutputPath(path) {
...
+  // To keep things "official" we follow the CSV specification (RFC4180)
+  // The document describes how to deal with escaping commas and quotes etc.
+  const CRLF = '\r\n';
+  const separator = ',';
+  /** @param {!string} value @returns {string} */
+  const escape = (value) => `"${value.replace(/"/g, '""')}"`;
+
+  // Map every audit to its corresponding category
+  // which can be queried when needed
+  const map = {};
```
Comment (patrickhulce) [inline lighthouse-cli/printer.js]: nit: let's give this a more descriptive name `auditIdToCategoryNameMap`?

File: lighthouse-cli/printer.js [hunk 2]
```diff
@@ -32,6 +34,48 @@ function checkOutputPath(path) {
...
+/**
+ * Converts the results to a CSV formatted string
+ * Each row describes the result of 1 audit with
+ *  - the name of the category the audit belongs to
+ *  - the name of the audit
+ *  - a description of the audit
+ *  - the score type that is used for the audit
+ *  - the score value of the audit
+ * 
+ * @param {!LH.Results} results
```
Comment (patrickhulce) [inline lighthouse-cli/printer.js]: nit: we don't need the `!` anymore

File: lighthouse-cli/printer.js [hunk 3]
```diff
@@ -32,6 +34,48 @@ function checkOutputPath(path) {
...
+ * @returns {string}
+ */
+function toCSVReport(results) {
+  // To keep things "official" we follow the CSV specification (RFC4180)
+  // The document describes how to deal with escaping commas and quotes etc.
+  const CRLF = '\r\n';
+  const separator = ',';
+  /** @param {!string} value @returns {string} */
```
Comment (patrickhulce) [inline lighthouse-cli/printer.js]: same `!`

File: package.json [hunk 4]
```diff
@@ -92,6 +92,7 @@
     "chrome-devtools-frontend": "1.0.422034",
     "chrome-launcher": "^0.10.2",
     "configstore": "^3.1.1",
+    "csv-validator": "^0.0.3",
```
Comment (patrickhulce) [inline package.json]: let's move this to devDependencies

File: lighthouse-cli/test/cli/printer-test.js [hunk 5]
```diff
@@ -35,6 +36,22 @@ describe('Printer', () => {
...
+      category: '',
+      name: '',
+      description: '',
+      type: '',
+      score: 42,
+    };
+    return Printer
+      .write(sampleResults, mode, path)
+      .then(_ => csv(path, headers).catch(err => assert.fail(err)))
+      .then(_ => fs.unlinkSync(path));
```
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-cli/test/cli/printer-test.js]: ah true. i was hoping `after` could be scoped to that test. whoops.

okay in that case, nevermind. :)

**Comment by paulirish** [comment]
wfm. over to patrick.

File: lighthouse-cli/printer.js [hunk 6]
```diff
@@ -32,6 +34,48 @@ function checkOutputPath(path) {
...
+  // which can be queried when needed
+  const auditCategories = {};
+  results.reportCategories.forEach(category => {
+    category.audits.forEach(audit => auditCategories[audit.id] = category.name);
+  });
+
+  // Possible TODO: tightly couple headers and row values
+  // This would make it easier to include new / other row values
+  const header = ['category', 'name', 'title', 'type', 'score'];
+  const table = Object
```
[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-cli/printer.js]: These arrays are so small it really comes down to style, though.

**Comment by patrickhulce** [comment]
ah yeah async/await cleans this up a lot! let's do it 👍 :)

**Comment by brendankenny** [comment]
appveyor errors should be fixed by #4933, so good to go. Thanks for the PR!

**Pull request merged**